chainindex/reconcile.go

```diff
@@ -0,0 +1,211 @@
+package chainindex
+
+import (
+	"context"
+	"database/sql"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+// ReconcileWithChain ensures that the index is consistent with the current chain state.
+// It performs the following steps:
+//  1. Checks if the index is empty. If so, it returns immediately as there's nothing to reconcile.
+//  2. Finds the lowest non-reverted height in the index.
+//  3. Walks backwards from the current chain head until it finds a tipset that exists
+//     in the index and is not marked as reverted.
+//  4. Sets a boundary epoch just above this found tipset.
+//  5. Marks all tipsets above this boundary as reverted, ensuring consistency with the current chain state.
+//  6. Applies all missing un-indexed tipsets starting from the last matching tipset b/w index and canonical chain
+//     to the current chain head.
+//
+// This function is crucial for maintaining index integrity, especially after chain reorgs.
+// It ensures that the index accurately reflects the current state of the blockchain.
+func (si *SqliteIndexer) ReconcileWithChain(ctx context.Context, head *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		si.closeLk.RUnlock()
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	if head == nil {
+		return nil
+	}
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		var hasTipset bool
+		err := tx.StmtContext(ctx, si.isTipsetMessageNonEmptyStmt).QueryRowContext(ctx).Scan(&hasTipset)
+		if err != nil {
+			return xerrors.Errorf("failed to check if tipset message is empty: %w", err)
+		}
+
+		isIndexEmpty := !hasTipset
+		if isIndexEmpty && !si.reconcileEmptyIndex {
+			log.Info("Chain index is empty and reconcileEmptyIndex is disabled; skipping reconciliation")
+			return nil
+		}
+
+		if isIndexEmpty {
+			log.Info("Chain index is empty; backfilling from head")
+			return si.backfillEmptyIndex(ctx, tx, head)
+		}
+
+		// Find the minimum applied tipset in the index; this will mark the absolute min height of the reconciliation walk
+		var reconciliationEpoch abi.ChainEpoch
+		row := tx.StmtContext(ctx, si.getMinNonRevertedHeightStmt).QueryRowContext(ctx)
+		if err := row.Scan(&reconciliationEpoch); err != nil {
+			return xerrors.Errorf("failed to scan minimum non-reverted height: %w", err)
+		}
```

Comment:
`isTipsetMessageNonEmptyStmt` doesn't care about reverted status, but we might have entries but they are all reverted. What's the behaviour going to be in that case? `SELECT MIN(height) FROM tipset_message WHERE reverted = 0` - will we error on this case or get a `reconciliationEpoch` of `0`? What do we want to happen in this case?

Reply:
@rvagg good catch, this would have thrown an exception as we're not using a nullable int here.
I've changed this code so that the reconciliation epoch is set to 0 if the index is non-empty but only contains reverted tipsets. That fixes this.

Reply:
k, trusting you on the sqlite behaviour on this one because I don't know without testing it myself